#include "mainwindow.h"
#include "ui_mainwindow.h"
#include "Manager.h"
#include "Cashier.h"

#include <QHeaderView>
#include <QStandardItem>
#include <QVariant>
#include <QPropertyAnimation>
#include <QMessageBox>
#include <QDebug>
#include <QInputDialog>
#include <QLabel>
#include <algorithm>

#include <qstring.h>
#include <qlineedit.h>
#include <qpushbutton.h>
#include <qlabel.h>
#include <qtableview.h>

#include <QGridLayout>
#include <QVBoxLayout>
#include <QFrame>
#include <QSize>

#include "BillItem.h"
#include "Exceptions.h"
#include "Payment.h"
#include "Customer.h"
#include "HashTable.h"

#include "AddProductToStore.h"
#include "ThongKe.h"
#include "AddCustomerToStore.h"
#include "ManageInventory.h"


MainWindow::MainWindow(User* user, Store* storePtr, QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindow), modelTable(nullptr), curTableProduct(0), store(storePtr), currentUser(user), currentSortCriteria(SORT_DEFAULT)
{
    ui->setupUi(this);
    setupTable();
    setupHoaDonTable();
    setupSortComboBox();
    ui->frameMenu->hide();

    connect(ui->ToanBo, &QPushButton::clicked, this, [this]() {
        ui->ToanBo->setChecked(true);
        ui->DoAn->setChecked(false);
        ui->ThucUong->setChecked(false);
        ui->DoGiaDung->setChecked(false);
        on_ToanBo_clicked();
    });

    connect(ui->DoAn, &QPushButton::clicked, this, [this]() {
        ui->ToanBo->setChecked(false);
        ui->DoAn->setChecked(true);
        ui->ThucUong->setChecked(false);
        ui->DoGiaDung->setChecked(false);
        on_DoAn_clicked();
    });

    connect(ui->ThucUong, &QPushButton::clicked, this, [this]() {
        ui->ToanBo->setChecked(false);
        ui->DoAn->setChecked(false);
        ui->ThucUong->setChecked(true);
        ui->DoGiaDung->setChecked(false);
        on_ThucUong_clicked();
    });

    connect(ui->DoGiaDung, &QPushButton::clicked, this, [this]() {
        ui->ToanBo->setChecked(false);
        ui->DoAn->setChecked(false);
        ui->ThucUong->setChecked(false);
        ui->DoGiaDung->setChecked(true);
        on_DoGiaDung_clicked();
    });

    connect(ui->btnToggleMenu, &QPushButton::clicked, this, &MainWindow::onToggleMenuClicked);
    connect(ui->btnCancelOrder, &QPushButton::clicked, this, &MainWindow::onCancelOrderClicked);
    connect(ui->tableViewProduct, &QTableView::doubleClicked, this, &MainWindow::onAddSanPham);
    connect(ui->SearchText, &QLineEdit::returnPressed, this, &MainWindow::on_BtnSearch_clicked);
    connect(ui->tableViewOrder, &QTableView::doubleClicked, this, &MainWindow::onRemoveSanPhamDoubleClicked);
    connect(ui->txtSearchCustomer, &QLineEdit::returnPressed, this, &MainWindow::onTimKhachPressed);
    connect(ui->txtSearchPhoneCustomer, &QLineEdit::returnPressed, this, &MainWindow::onTimKhachPressed);
    connect(ui->btnDungDiem, &QPushButton::clicked, this, &MainWindow::onDungDiemClicked);
    connect(ui->ExportOrder, &QPushButton::clicked, this, &MainWindow::onXuatHoaDonClicked);
    connect(ui->btnBack, &QPushButton::clicked, this, &MainWindow::onQuayLaiClicked);
    connect(ui->btnCash, &QPushButton::clicked, this, &MainWindow::onThanhToanTienMatClicked);
    connect(ui->btnCard, &QPushButton::clicked, this, &MainWindow::onThanhToanTheClicked);
    connect(ui->btnExport, &QPushButton::clicked, this, &MainWindow::onThanhToanClicked);

    loadAndSortProducts(0);
    currentBill = new Bill(nullptr, "", currentUser);

    ui->stackedWidgeOrder->setCurrentIndex(curTableProduct);

    applyPermissions();
    QHeaderView* header = ui->tableViewProduct->horizontalHeader();
    header->setMaximumSectionSize(500);
    header->setMinimumSectionSize(100);
    header->setStretchLastSection(false);
    header->setSectionResizeMode(0, QHeaderView::ResizeToContents);
    header->setSectionResizeMode(1, QHeaderView::Interactive);
    header->resizeSection(1, 300);
    for (int i = 2; i <= 6; i++) {
        header->setSectionResizeMode(i, QHeaderView::Stretch);
    }
    header->setSectionResizeMode(7, QHeaderView::ResizeToContents);
}

void MainWindow::setupSortComboBox()
{
    ui->sortComboBox->setMinimumHeight(45);
    ui->sortComboBox->setMinimumWidth(200);

    ui->sortComboBox->addItem("M·∫∑c ƒë·ªãnh", SORT_DEFAULT);
    ui->sortComboBox->addItem("Gi√°: TƒÉng d·∫ßn", SORT_PRICE_ASC);
    ui->sortComboBox->addItem("Gi√°: Gi·∫£m d·∫ßn", SORT_PRICE_DESC);
    ui->sortComboBox->addItem("S·ªë l∆∞·ª£ng: TƒÉng d·∫ßn", SORT_QUANTITY_ASC);
    ui->sortComboBox->addItem("S·ªë l∆∞·ª£ng: Gi·∫£m d·∫ßn", SORT_QUANTITY_DESC);
    ui->sortComboBox->addItem("H·∫°n SD: G·∫ßn nh·∫•t", SORT_EXPIRY_ASC);
    ui->sortComboBox->addItem("H·∫°n SD: Xa nh·∫•t", SORT_EXPIRY_DESC);
    ui->sortComboBox->addItem("Th·ªÉ t√≠ch: TƒÉng d·∫ßn", SORT_VOLUME_ASC);
    ui->sortComboBox->addItem("Th·ªÉ t√≠ch: Gi·∫£m d·∫ßn", SORT_VOLUME_DESC);
    ui->sortComboBox->addItem("B·∫£o h√†nh: TƒÉng d·∫ßn", SORT_WARRANTY_ASC);
    ui->sortComboBox->addItem("B·∫£o h√†nh: Gi·∫£m d·∫ßn", SORT_WARRANTY_DESC);

    connect(ui->sortComboBox, QOverload<int>::of(&QComboBox::currentIndexChanged),
            this, &MainWindow::onSortCriteriaChanged);
}

void MainWindow::onSortCriteriaChanged(int index)
{
    currentSortCriteria = static_cast<SortCriteria>(ui->sortComboBox->itemData(index).toInt());
    loadAndSortProducts(curTableProduct);
}

void MainWindow::applyPermissions()
{
    bool isAdmin = (dynamic_cast<Manager*>(currentUser) != nullptr);
    bool isCashier = (dynamic_cast<Cashier*>(currentUser) != nullptr);
    ui->ThemHang->setVisible(isAdmin);
    ui->ThongKe->setVisible(isAdmin || isCashier);
    ui->KhachHang->setVisible(isAdmin || isCashier);

    QString userInfo = QString("%1 (%2)").arg(currentUser->getName()).arg(isAdmin ? "Qu·∫£n l√Ω" : "Nh√¢n vi√™n");
    ui->lblStoreName->setText(QString("C·ª¨A H√ÄNG T·∫†P H√ìA\n%1").arg(userInfo));
}

void MainWindow::onToggleMenuClicked()
{
    bool isVisible = ui->frameMenu->isVisible();
    ui->frameMenu->setVisible(!isVisible);
}

void MainWindow::onCancelOrderClicked()
{
    if(currentBill == nullptr)
        return;
    if (currentBill->getCustomer() == nullptr &&  currentBill->getItems().empty())
    {
        ui->txtSearchCustomer->clear();
        ui->txtSearchPhoneCustomer->clear();
        ui->lblTenKhach->setText("Kh√°ch L·∫ª");
        return;
    }

    QMessageBox::StandardButton reply;
    reply = QMessageBox::question(this, "X√°c nh·∫≠n h·ªßy",
                                  "B·∫°n c√≥ ch·∫Øc mu·ªën h·ªßy h√≥a ƒë∆°n n√†y?\nT·∫•t c·∫£ s·∫£n ph·∫©m s·∫Ω ƒë∆∞·ª£c tr·∫£ v·ªÅ kho.",
                                  QMessageBox::Yes | QMessageBox::No);

    if (reply == QMessageBox::Yes)
    {
        resetHoaDon();
        updateHoaDonView();
        QMessageBox::information(this, "Th√†nh c√¥ng", "ƒê√£ h·ªßy h√≥a ƒë∆°n v√† tr·∫£ h√†ng v·ªÅ kho.");
    }
}

MainWindow::~MainWindow()
{
    delete ui;
    delete currentBill;
}

void MainWindow::setupTable()
{
    modelTable = new QStandardItemModel(this);
    modelTable->setColumnCount(8);
    modelTable->setHeaderData(0, Qt::Horizontal, "M√£ s·∫£n ph·∫©m");
    modelTable->setHeaderData(1, Qt::Horizontal, "T√™n s·∫£n ph·∫©m");
    modelTable->setHeaderData(2, Qt::Horizontal, "Lo·∫°i s·∫£n ph·∫©m");
    modelTable->setHeaderData(3, Qt::Horizontal, "Gi√° s·∫£n ph·∫©m");
    modelTable->setHeaderData(4, Qt::Horizontal, "S·ªë l∆∞·ª£ng");
    modelTable->setHeaderData(5, Qt::Horizontal, "Th·ªÉ t√≠ch");
    modelTable->setHeaderData(6, Qt::Horizontal, "H·∫°n s·ª≠ d·ª•ng");
    modelTable->setHeaderData(7, Qt::Horizontal, "Th·ªùi h·∫°n b·∫£o h√†nh");
    ui->tableViewProduct->setModel(modelTable);
    ui->tableViewProduct->setSelectionBehavior(QAbstractItemView::SelectRows);
    ui->tableViewProduct->setSelectionMode(QAbstractItemView::SingleSelection);
    ui->tableViewProduct->setEditTriggers(QAbstractItemView::NoEditTriggers);
}

void MainWindow::setupHoaDonTable()
{
    modelHoaDon = new QStandardItemModel(this);
    modelHoaDon->setColumnCount(3);
    modelHoaDon->setHeaderData(0, Qt::Horizontal, "T√™n SP");
    modelHoaDon->setHeaderData(1, Qt::Horizontal, "SL");
    modelHoaDon->setHeaderData(2, Qt::Horizontal, "Th√†nh ti·ªÅn");
    ui->tableViewOrder->setModel(modelHoaDon);
    ui->tableViewOrder->horizontalHeader()->setSectionResizeMode(QHeaderView::Stretch);
}

void MainWindow::setupLastBill()
{
    modelLastBill = new QStandardItemModel(this);
    modelLastBill->setColumnCount(3);
    modelLastBill->setHeaderData(0, Qt::Horizontal, "T√™n SP");
    modelLastBill->setHeaderData(1, Qt::Horizontal, "SL");
    modelLastBill->setHeaderData(2, Qt::Horizontal, "Th√†nh ti·ªÅn");
    ui->tableLastBill->setModel(modelLastBill);
    ui->tableLastBill->horizontalHeader()->setSectionResizeMode(QHeaderView::Stretch);
}

void MainWindow::updateHoaDonView()
{
    qDebug() << 1 << '\n';
    modelHoaDon->removeRows(0, modelHoaDon->rowCount());

    if(currentBill)
    {
        const auto& items = currentBill->getItems();
        for (const BillItem& item : items)
        {
            QList<QStandardItem*> row;
            row << new QStandardItem(item.getProduct()->getName());
            row << new QStandardItem(QString::number(item.getQuantity()));
            row << new QStandardItem(QString::number(item.getLineTotal()));
            modelHoaDon->appendRow(row);
        }

        double subTotal = currentBill->getSubTotal();
        double finalTotal = currentBill->getTotal();
        if(currentBill->getCustomer())
        {
            ui->lblTenKhach->setText(currentBill->getCustomer()->getName());
            ui->lblDiemKhach->setText(QString("ƒêi·ªÉm T√≠ch L≈©y: %1").arg(currentBill->getCustomer()->getPoints()));
        }
        ui->TotalBefore->setText(QString("T·ªïng ti·ªÅn ban ƒë·∫ßu: %1").arg(subTotal));
        ui->TotalAfter->setText(QString("T·ªïng ti·ªÅn thanh to√°n: %1").arg(finalTotal));
    }
}

void MainWindow::updateLastBillView()
{
    modelLastBill->removeRows(0, modelLastBill->rowCount());
    const auto& items = currentBill->getItems();
    for (const BillItem& item : items)
    {
        QList<QStandardItem*> row;
        row << new QStandardItem(item.getProduct()->getName());
        row << new QStandardItem(QString::number(item.getQuantity()));
        row << new QStandardItem(QString::number(item.getLineTotal()));
        modelLastBill->appendRow(row);
    }

    double subTotal = currentBill->getSubTotal();
    double finalTotal = currentBill->getTotal();
    ui->TotalBefore_2->setText(QString("T·ªïng ti·ªÅn ban ƒë·∫ßu: %1").arg(subTotal));
    ui->TotalAfter_2->setText(QString("T·ªïng ti·ªÅn thanh to√°n: %1").arg(finalTotal));
}

void MainWindow::resetHoaDon()
{
    ui->txtSearchCustomer->clear();
    ui->txtSearchPhoneCustomer->clear();
    ui->btnDungDiem->setEnabled(false);
    ui->stackedWidgeOrder->setCurrentIndex(0);
    ui->lblTenKhach->setText("Kh√°ch l·∫ª");
    ui->lblDiemKhach->setText("");
    ui->TotalBefore->setText("");
    ui->TotalAfter->setText("");

    // ‚úÖ TR·∫¢ H√ÄNG V·ªÄ KHO khi h·ªßy ƒë∆°n
    const std::vector<BillItem>& billitems = currentBill->getItems();
    for(const BillItem& item : billitems)
    {
        // üõ°Ô∏è FIX: D√πng ID thay v√¨ Name (tr√°nh tr·∫£ nh·∫ßm s·∫£n ph·∫©m c√πng t√™n)
        Product* p = store->findProductById(item.getProduct()->getId());
        if(p) {
            p->setQuantity(p->getQuantity() + item.getQuantity());
        }
    }

    Customer* c = currentBill->getCustomer();
    if(c && currentBill->getCheck())
        c->setPoints(c->getPoints() + 100000);

    currentBill = nullptr;
}

void MainWindow::loadAndSortProducts(int typeFilter)
{
    std::vector<Product*> products;

    store->forEachProduct([&](const QString&, Product* p) {
        if (!p) return;

        bool shouldInclude = false;
        Food* f = dynamic_cast<Food*>(p);
        Beverage* b = dynamic_cast<Beverage*>(p);
        HouseholdItem* h = dynamic_cast<HouseholdItem*>(p);

        if (typeFilter == 0) {
            shouldInclude = true;
        } else if (typeFilter == 1 && f) {
            shouldInclude = true;
        } else if (typeFilter == 2 && b) {
            shouldInclude = true;
        } else if (typeFilter == 3 && h) {
            shouldInclude = true;
        }

        if (shouldInclude) {
            products.push_back(p);
        }
    });

    applySortingAndFiltering(products);

    modelTable->removeRows(0, modelTable->rowCount());

    for (Product* p : products) {
        if (!p) continue;

        Food* f = dynamic_cast<Food*>(p);
        Beverage* b = dynamic_cast<Beverage*>(p);
        HouseholdItem* h = dynamic_cast<HouseholdItem*>(p);

        QList<QStandardItem*> row;
        row << new QStandardItem(p->getId());
        row << new QStandardItem(p->getName());

        QString typeName;
        if (f) typeName = "ƒê·ªì ƒÉn";
        else if (b) typeName = "Th·ª©c u·ªëng";
        else if (h) typeName = "ƒê·ªì gia d·ª•ng";
        else typeName = "Kh√°c";
        row << new QStandardItem(typeName);

        row << new QStandardItem(QString::number(p->calcFinalPrice()));
        row << new QStandardItem(QString::number(p->getQuantity()));
        row << new QStandardItem(b ? QString::number(b->getVolume()) : "");

        if(f)
            row << new QStandardItem(f->getExpiryDate());
        else if(b)
            row << new QStandardItem(b->getExpiryDate());
        else
            row << new QStandardItem("");

        row << new QStandardItem(h ? QString::number(h->getWarrantyMonths()) : "");
        modelTable->appendRow(row);
    }
}

void MainWindow::applySortingAndFiltering(std::vector<Product*>& products)
{
    switch (currentSortCriteria) {
    case SORT_DEFAULT:
        break;

    case SORT_PRICE_ASC:
        std::sort(products.begin(), products.end(), comparePriceAsc);
        break;

    case SORT_PRICE_DESC:
        std::sort(products.begin(), products.end(), comparePriceDesc);
        break;

    case SORT_QUANTITY_ASC:
        std::sort(products.begin(), products.end(), compareQuantityAsc);
        break;

    case SORT_QUANTITY_DESC:
        std::sort(products.begin(), products.end(), compareQuantityDesc);
        break;

    case SORT_EXPIRY_ASC:
    case SORT_EXPIRY_DESC:
        products.erase(
            std::remove_if(products.begin(), products.end(), [](Product* p) {
                return dynamic_cast<HouseholdItem*>(p);
            }),
            products.end()
            );
        if(currentSortCriteria == SORT_EXPIRY_ASC)
            std::sort(products.begin(), products.end(), compareExpiryAsc);
        else
            std::sort(products.begin(), products.end(), compareExpiryDesc);
        break;
    case SORT_VOLUME_ASC:
    case SORT_VOLUME_DESC:
        products.erase(
            std::remove_if(products.begin(), products.end(), [](Product* p) {
                return dynamic_cast<Beverage*>(p) == nullptr;
            }),
            products.end()
            );
        if (currentSortCriteria == SORT_VOLUME_ASC)
            std::sort(products.begin(), products.end(), compareVolumeAsc);
        else
            std::sort(products.begin(), products.end(), compareVolumeDesc);
        break;

    case SORT_WARRANTY_ASC:
    case SORT_WARRANTY_DESC:
        products.erase(
            std::remove_if(products.begin(), products.end(), [](Product* p) {
                return dynamic_cast<HouseholdItem*>(p) == nullptr;
            }),
            products.end()
            );
        if (currentSortCriteria == SORT_WARRANTY_ASC)
            std::sort(products.begin(), products.end(), compareWarrantyAsc);
        else
            std::sort(products.begin(), products.end(), compareWarrantyDesc);
        break;
    }
}

bool MainWindow::comparePriceAsc(Product* a, Product* b)
{
    if (!a || !b) return false;
    return a->calcFinalPrice() < b->calcFinalPrice();
}

bool MainWindow::comparePriceDesc(Product* a, Product* b)
{
    if (!a || !b) return false;
    return a->calcFinalPrice() > b->calcFinalPrice();
}

bool MainWindow::compareQuantityAsc(Product* a, Product* b)
{
    if (!a || !b) return false;
    return a->getQuantity() < b->getQuantity();
}

bool MainWindow::compareQuantityDesc(Product* a, Product* b)
{
    if (!a || !b) return false;
    return a->getQuantity() > b->getQuantity();
}

bool MainWindow::compareExpiryAsc(Product* a, Product* b)
{
    if (!a || !b) return false;

    QDate d1 = getProductDate(a);
    QDate d2 = getProductDate(b);

    if (!d1.isValid() && !d2.isValid()) return false;
    if (!d1.isValid()) return false;
    if (!d2.isValid()) return false;

    return d1 < d2;
}

bool MainWindow::compareExpiryDesc(Product* a, Product* b)
{
    if (!a || !b) return false;

    QDate d1 = getProductDate(a);
    QDate d2 = getProductDate(b);

    if (!d1.isValid() && !d2.isValid()) return false;
    if (!d1.isValid()) return false;
    if (!d2.isValid()) return true;

    return d1 > d2;
}

bool MainWindow::compareVolumeAsc(Product* a, Product* b)
{
    if (!a || !b) return false;

    Beverage* b1 = dynamic_cast<Beverage*>(a);
    Beverage* b2 = dynamic_cast<Beverage*>(b);

    if (!b1 || !b2) return false;

    return b1->getVolume() < b2->getVolume();
}

bool MainWindow::compareVolumeDesc(Product* a, Product* b)
{
    if (!a || !b) return false;

    Beverage* b1 = dynamic_cast<Beverage*>(a);
    Beverage* b2 = dynamic_cast<Beverage*>(b);

    if (!b1 || !b2) return false;

    return b1->getVolume() > b2->getVolume();
}

bool MainWindow::compareWarrantyAsc(Product* a, Product* b)
{
    if (!a || !b) return false;

    HouseholdItem* h1 = dynamic_cast<HouseholdItem*>(a);
    HouseholdItem* h2 = dynamic_cast<HouseholdItem*>(b);

    if (!h1 || !h2) return false;

    return h1->getWarrantyMonths() < h2->getWarrantyMonths();
}

bool MainWindow::compareWarrantyDesc(Product* a, Product* b)
{
    if (!a || !b) return false;

    HouseholdItem* h1 = dynamic_cast<HouseholdItem*>(a);
    HouseholdItem* h2 = dynamic_cast<HouseholdItem*>(b);

    if (!h1 || !h2) return false;

    return h1->getWarrantyMonths() > h2->getWarrantyMonths();
}

QDate MainWindow::getProductDate(Product* p)
{
    if (!p) return QDate();

    QString dateStr;
    if (auto f = dynamic_cast<Food*>(p))
        dateStr = f->getExpiryDate();
    else if (auto b = dynamic_cast<Beverage*>(p))
        dateStr = b->getExpiryDate();

    if (dateStr.isEmpty()) return QDate();
    return QDate::fromString(dateStr, "dd/MM/yyyy");
}

void MainWindow::loadProductsFromStore(int typeFilter)
{
    loadAndSortProducts(typeFilter);
}

void MainWindow::loadProductsFromStoreWithKeyWord(const QString &keyword)
{
    QString kw = keyword.trimmed();
    if (kw.isEmpty())
    {
        loadAndSortProducts(0);
        return;
    }

    modelTable->removeRows(0, modelTable->rowCount());

    store->forEachProductByName(kw, [&](const QString&, Product* p) {
        if (!p) return;

        Food* f = dynamic_cast<Food*>(p);
        Beverage* b = dynamic_cast<Beverage*>(p);
        HouseholdItem* h = dynamic_cast<HouseholdItem*>(p);

        QList<QStandardItem*> row;
        row << new QStandardItem(p->getId());
        row << new QStandardItem(p->getName());

        QString typeName;
        if (f) typeName = "ƒê·ªì ƒÉn";
        else if(b) typeName = "Th·ª©c u·ªëng";
        else if (h) typeName = "ƒê·ªì gia d·ª•ng";
        row << new QStandardItem(typeName);

        row << new QStandardItem(QString::number(p->calcFinalPrice()));
        row << new QStandardItem(QString::number(p->getQuantity()));
        row << new QStandardItem(b ? QString::number(b->getVolume()) : "");

        if(f)
            row << new QStandardItem(f->getExpiryDate());
        else if(b)
            row << new QStandardItem(b->getExpiryDate());
        else
            row << new QStandardItem("");

        row << new QStandardItem(h ? QString::number(h->getWarrantyMonths()) : "");
        modelTable->appendRow(row);
    });
}

void MainWindow::on_ToanBo_clicked()
{
    loadAndSortProducts(0);
    curTableProduct = 0;
}

void MainWindow::on_DoAn_clicked()
{
    loadAndSortProducts(1);
    curTableProduct = 1;
}

void MainWindow::on_ThucUong_clicked()
{
    loadAndSortProducts(2);
    curTableProduct = 2;
}

void MainWindow::on_DoGiaDung_clicked()
{
    loadAndSortProducts(3);
    curTableProduct = 3;
}

void MainWindow::on_BtnSearch_clicked()
{
    loadProductsFromStoreWithKeyWord(ui->SearchText->text());
}

void MainWindow::onAddSanPham(const QModelIndex &index)
{
    auto v2 = ui->stackedWidgeOrder->currentIndex();
    if(v2 == 1)
        return;

    if(!index.isValid()) {
        QMessageBox::warning(this, "L·ªói", "Vui l√≤ng ch·ªçn s·∫£n ph·∫©m h·ª£p l·ªá.");
        return;
    }

    QString Id = modelTable->item(index.row(), 0)->text();
    Product* p = store->findProductById(Id);

    if(!p) {
        QMessageBox::warning(this, "L·ªói", "Kh√¥ng t√¨m th·∫•y s·∫£n ph·∫©m.");
        return;
    }

    int maxStock = p->getQuantity();
    if (maxStock <= 0)
    {
        QMessageBox::warning(this, "H·∫øt h√†ng", "S·∫£n ph·∫©m n√†y ƒë√£ h·∫øt h√†ng.");
        return;
    }

    bool ok;
    int quantityToAdd = QInputDialog::getInt(
        this,
        "Nh·∫≠p s·ªë l∆∞·ª£ng",
        QString("Nh·∫≠p s·ªë l∆∞·ª£ng cho %1:").arg(p->getName()), 1, 1, maxStock, 1, &ok);

    if(ok)
    {
        if(currentBill == nullptr)
            currentBill = new Bill(nullptr, "", currentUser);
        currentBill->addItem(p, quantityToAdd);
        ui->stackedWidgeOrder->setCurrentIndex(0);
        loadAndSortProducts(curTableProduct);
        updateHoaDonView();
    }
}

void MainWindow::onRemoveSanPhamDoubleClicked(const QModelIndex &index)
{
    QString name = modelHoaDon->item(index.row(), 0)->text();
    Product* p = store->findProductByName(name);
    currentBill->removeItem(p);
    updateHoaDonView();
}

void MainWindow::onTimKhachPressed()
{
    QString phone = ui->txtSearchPhoneCustomer->text().trimmed();
    QString name = ui->txtSearchCustomer->text().trimmed();
    if (phone.isEmpty() || name.isEmpty())
        return;

    Customer* c = store->findCustomerByPhone(phone);
    if (c == nullptr)
    {
        currentBill->setCustomer(nullptr);
        ui->lblTenKhach->setText("Kh√¥ng t√¨m th·∫•y!");
        ui->lblTenKhach->setStyleSheet("color: red;");
        ui->lblDiemKhach->setText("");
        ui->btnDungDiem->setEnabled(false);
        return;
    }
    if (c->getName().toLower() != name.toLower())
    {
        ui->lblTenKhach->setText("Th√¥ng tin kh√¥ng kh·ªõp!");
        ui->lblTenKhach->setStyleSheet("color: red;");
        ui->lblDiemKhach->setText("");
        ui->btnDungDiem->setEnabled(false);
        return;
    }
    if(currentBill == nullptr)
        currentBill = new Bill(nullptr, "", currentUser);
    currentBill->setCustomer(c);
    ui->lblTenKhach->setText(c->getName());
    ui->lblDiemKhach->setText(QString("ƒêi·ªÉm T√≠ch L≈©y: %1").arg(c->getPoints()));

    if (c->getPoints() >= 100000)
        ui->btnDungDiem->setEnabled(true);
    else
        ui->btnDungDiem->setEnabled(false);
}

void MainWindow::onDungDiemClicked()
{
    bool success = currentBill->applyPointsDiscount(100000);
    currentBill->setCheck(true);
    if (success)
    {
        updateHoaDonView();
        ui->btnDungDiem->setEnabled(false);
        QMessageBox::information(this, "Th√†nh c√¥ng", "ƒê√£ √°p d·ª•ng gi·∫£m gi√° 2% v√† tr·ª´ 100.000 ƒëi·ªÉm.");
    }
    else
        QMessageBox::warning(this, "L·ªói", "Kh√¥ng ƒë·ªß ƒëi·ªÉm ho·∫∑c ƒë√£ √°p d·ª•ng gi·∫£m gi√°.");
}

void MainWindow::onXuatHoaDonClicked()
{
    if (currentBill->getItems().empty())
    {
        QMessageBox::warning(this, "H√≥a ƒë∆°n tr·ªëng", "Vui l√≤ng ch·ªçn s·∫£n ph·∫©m.");
        return;
    }
    ui->stackedWidgeOrder->setCurrentIndex(1);
    setupLastBill();
    updateLastBillView();
}

void MainWindow::onQuayLaiClicked()
{
    ui->stackedWidgeOrder->setCurrentIndex(0);
}

void MainWindow::onThanhToanTienMatClicked()
{
    currentBill->setPayment(new CashPayment());
    QMessageBox::information(this, "Th√¥ng B√°o: ", "ƒê√£ Ch·ªçn ph∆∞∆°ng th·ª©c thanh to√°n l√† ti·ªÅn m·∫∑t");
}

void MainWindow::onThanhToanTheClicked()
{
    currentBill->setPayment(new CardPayment());
    QMessageBox::information(this, "Th√¥ng B√°o: ", "ƒê√£ Ch·ªçn ph∆∞∆°ng th·ª©c thanh to√°n l√† qu·∫πt th·∫ª");
}

void MainWindow::onThanhToanClicked()
{
    if(currentBill->getPayment() == nullptr)
    {
        QMessageBox::information(this, "Th√¥ng B√°o: ", "Vui l√≤ng ch·ªçn ph∆∞∆°ng th·ª©c thanh to√°n tr∆∞·ªõc khi thanh to√°n");
        return;
    }
    finalizeThanhToan(currentBill->getPayment()->getMethodName());
    ui->stackedWidgeOrder->setCurrentIndex(0);
    currentBill = new Bill(nullptr, "", currentUser);
    updateHoaDonView();
    loadAndSortProducts(curTableProduct);
}

// ‚úÖ BLUE TEAM FIX: Atomic Transaction v·ªõi Rollback Protection
void MainWindow::finalizeThanhToan(const QString& paymentMethod)
{
    // üõ°Ô∏è B∆Ø·ªöC 1: RE-CHECK STOCK (ƒê·ªÅ ph√≤ng race condition)
    const std::vector<BillItem>& billItems = currentBill->getItems();
    for (const BillItem& item : billItems)
    {
        Product* p = item.getProduct();
        int needed = item.getQuantity();
        int available = p->getQuantity();

        if (available < 0) // Kho ƒë√£ b·ªã √¢m
        {
            QMessageBox::critical(this, "L·ªói Nghi√™m Tr·ªçng",
                QString("S·∫£n ph·∫©m '%1' c√≥ t·ªìn kho b·∫•t th∆∞·ªùng (%2). H·ªßy giao d·ªãch!")
                .arg(p->getName()).arg(available));
            
            // Rollback: Tr·∫£ h√†ng v·ªÅ kho
            for (const BillItem& rollbackItem : billItems)
            {
                Product* rp = rollbackItem.getProduct();
                rp->setQuantity(rp->getQuantity() + rollbackItem.getQuantity());
            }
            return; // D·ª™NG GIAO D·ªäCH
        }
    }

    double finalTotal = currentBill->getTotal();
    store->addRevenue(finalTotal);

    Customer* c = currentBill->getCustomer();
    if (c != nullptr)
    {
        int pointsToAdd = static_cast<int>(finalTotal * 0.10);
        c->addPoints(pointsToAdd);
    }

    store->addBillToHistory(currentBill);

    QMessageBox::information(this, "Th√†nh c√¥ng",
        QString("ƒê√£ thanh to√°n %1 ƒë·ªìng b·∫±ng %2\n")
        .arg(finalTotal).arg(paymentMethod));
}

void MainWindow::on_ThemHang_clicked()
{
    AddProductToStore dialog(this);
    int result = dialog.exec();

    if (result == QDialog::Accepted)
    {
        QString type = dialog.getProductType();
        QString name = dialog.getName();
        double price = dialog.getPrice();              // Gi√° b√°n (t·ª± ƒë·ªông t√≠nh)
        double importPrice = dialog.getImportPrice();  // Gi√° g·ªëc
        double profitMargin = dialog.getProfitMargin();// % L·ª£i nhu·∫≠n
        int quantity = dialog.getQuantity();

        if (type == "ƒê·ªì ƒÉn")
        {
            QString expiry = dialog.getExpiryDate();
            QString id = store->generateFoodId();
            Product* p = new Food(id, name, price, quantity, expiry);
            p->setImportPrice(importPrice);
            p->setProfitMargin(profitMargin);
            store->addProduct(p);
        }
        else if (type == "Th·ª©c u·ªëng")
        {
            QString expiry = dialog.getExpiryDate();
            double volume = dialog.getVolume();
            QString id = store->generateBeverageId();
            Product* p = new Beverage(id, name, price, quantity, expiry, volume);
            p->setImportPrice(importPrice);
            p->setProfitMargin(profitMargin);
            store->addProduct(p);
        }
        else if (type == "ƒê·ªì gia d·ª•ng")
        {
            int warranty = dialog.getWarranty();
            QString id = store->generateHouseholdId();
            Product* p = new HouseholdItem(id, name, price, quantity, warranty);
            p->setImportPrice(importPrice);
            p->setProfitMargin(profitMargin);
            store->addProduct(p);
        }
        loadAndSortProducts(curTableProduct);
    }
}

void MainWindow::on_ThongKe_clicked()
{
    ThongKe dialog(store, this);
    dialog.exec();
}

void MainWindow::on_KhachHang_clicked()
{
    CustomerDialog dialog(store, this);
    dialog.exec();
}


void MainWindow::on_QuanLyKho_clicked()
{
    ManageInventory dialog(store, this);
    dialog.exec();
    loadAndSortProducts(curTableProduct);
}

